snuba/utils/streams/kafka.py

```diff
@@ -98,12 +171,50 @@ def poll(self, timeout: Optional[float] = None) -> Optional[KafkaMessage]:
             else:
                 raise ConsumerError(str(error))
 
-        return KafkaMessage(
+        result = KafkaMessage(
             TopicPartition(message.topic(), message.partition()),
             message.offset(),
             message.value(),
         )
 
+        self.__offsets[result.stream] = result.offset + 1
+
+        return result
+
+    def tell(self) -> Mapping[TopicPartition, int]:
+        if self.__state is ConsumerState.CLOSED:
+            raise RuntimeError("consumer already closed")
+
+        return self.__offsets
```

Comment:
What if another consumer is consuming on the same topic on the same consumer group ? Can that consumer group be assigned the same partition and thus advance the committed offset making this number wrong ? Or dues the broker guarantee a partition is assigned only to at most one consumer at a time ?

Reply:
> Or dues the broker guarantee a partition is assigned only to at most one consumer at a time ?

Yes, a consumer should have exclusive ownership over a partition, as long as it is still sending heartbeats to the broker and considered part of the consumer group. It's possible for the consumer itself to *think* that it still has partition ownership when it doesn't, but trying to commit will be rejected by the broker (this is why we see `KafkaError{​code=UNKNO­WN_MEMBER_­ID,​val=25,​str="Commi­t failed: Broker: Unknown member"}` when the replacer takes too long to rebalance and gets kicked out of the consumer group.)